arrows/vxl/pixel_feature_extractor.cxx

```diff
@@ -150,18 +191,50 @@ ::concatenate_images( std::vector< vil_image_view_base_sptr > filtered_images )
   return concatenated_planes;
 }
 
+// ----------------------------------------------------------------------------
+// Convert to a narrower type without wrapping
+template < typename out_t, typename in_t >
+vil_image_view< out_t >
+clamping_cast( vil_image_view< in_t > input_image )
+{
+  constexpr auto min_output_value =
+    static_cast< in_t >( std::numeric_limits< out_t >::min() );
+  constexpr auto max_output_value =
+    static_cast< in_t >( std::numeric_limits< out_t >::max() );
+
+  vil_clamp( input_image, input_image, min_output_value, max_output_value );
```

Comment:
While the improved clarity is certainly nice, I wonder if this isn't slower than the previous version.

Also, what happens if the output range is greater than the input range; do `{min,max}_output_value` get set reasonably in that case?

Reply:
Looking at the [source code](https://vxl.github.io/doc/release/core/vil/html/vil__clamp_8h_source.html), I don't think this would be much slower than applying the functor to every pixel, unless there's something I'm missing.

However, I see your point about the cast doing something weird. I'm actually not sure how to address this since `std::min/max` only are templated on one type. And if we accidentally cast to the smaller type before doing this we'd hit the same issue. I'd imagine there's a canonical solution but I haven't found it yet.

Reply:
Hmm, you're right. I think I was mainly thinking that there was an additional intermediate image, but it is in-place. It is probably still be slower due to cache thrashing, but I'm not sure the difference is enough to justify losing the gains in code simplification.

I haven't worked through how to write the code to deal with the range issue, but I think it is something like:
```c++
  using limit_t = promoted<in_t, out_t>::type;
  constexpr auto in_min = static_cast<limit_t>(limits<out_t>::min();
  // etc.
  if (in_min < out_min || in_max > out_max)
    // only do the clamping here, and the existing code is safe
```

Where this gets annoying is you may need branches for if `in_t` is floating-point, or one or both directions of sign conversions occurring. Maybe this is complicated enough it should be punted to a new PR, but we probably want to handle it, as it seems to me that the code in its current state may misbehave otherwise.

Or maybe I'm overthinking this, and the correct way to handle it is to convert the limits immediately to double, then test if we need to clamp, and only then cast back to in_t (which will theoretically be safe). Especially as I believe the comparison can be optimized away, so there is no performance penalty using `double` there.

Reply:
...I believe you still need to take the nearer-to-zero of both bounds before casting to `in_t`, though, as trying to clamp to `static_cast<int32_t>(limits<uint32_t>::max())` may still end badly.